service/src/main/java/bio/terra/cbas/controllers/RunSetsApiController.java

```diff
@@ -333,11 +294,29 @@ public ResponseEntity<AbortRunSetResponse> abortRunSet(UUID runSetId) {
       throw new ForbiddenException(SamService.WRITE_ACTION, SamService.RESOURCE_TYPE_WORKSPACE);
     }
 
+    // Get the run set associated with runSetId
+    RunSet runSet = runSetDao.getRunSet(runSetId);
+
+    // If RunSet is in Queued state, it means that CBAS is still processing the POST request in the
+    // async method. In this case, if the RunSet is set to Cancelling, it might leave the RunSet in
+    // weird state as it's possible that the background thread could still be processing Runs
+    // when the abort request came in and as a result some Runs that had been submitted to
+    // Cromwell got cancelled but some Runs that were still in process of being submitted do in
+    // fact get submitted to Cromwell and are started by Cromwell. And hence the RunSet might have
+    // some aborted Runs and some Runs in Running state.
+    if (runSet.status() == CbasRunSetStatus.QUEUED) {
+      String errorMessage =
+          "Run Set can't be aborted when it is Queued state as system might still be processing the request.";
+      return new ResponseEntity<>(
+          new AbortRunSetResponse().runSetId(runSetId).errors(errorMessage),
+          HttpStatus.BAD_REQUEST);
+    }
```

Comment:
This changes when a RunSet can be aborted. Since the submission happens in a different thread it is difficult to know how many Runs have been submitted to Cromwell and how many haven't. Hence the proposal here is to not allow abort requests when RunSet is in Queued state. I think when we move to a better async submissions design where CBAS has queues or scheduler to track the work it might be possible abort RunSet even in Queued state but it would be difficult in this iteration.